zerver/actions/message_flags.py

```diff
@@ -263,13 +263,14 @@ def do_update_message_flags(
     flagattr = getattr(UserMessage.flags, flag)
 
     with transaction.atomic(savepoint=False):
-        if flag == "read" and operation == "remove":
-            # We have an invariant that all stream messages marked as
-            # unread must be in streams the user is subscribed to.
+        if flag == "read":
+            # We have an invariant that all stream messages marked
+            # as read or unread must be in streams the user is
+            # subscribed to.
             #
-            # When marking as unread, we enforce this invariant by
-            # ignoring any messages in streams the user is not
-            # currently subscribed to.
+            # When marking as read or unread, we enforce this
+            # invariant by ignoring any messages in streams the
+            # user is not currently subscribed to.
```

Comment:
So thinking about this a bit more -- if a user ends up in a state where they have unread messages with existing UserMessage rows in an unsubscribed streams (which we know can happen in some race conditions) ... they will be displayed as such in the web app, and I think this code path would prevent the user from marking those as read and thus clearing the state. (I guess a user could still clear these via marking the whole stream as read?)

So I wonder whether in the case of marking-as-read, the most correct thing to do would be to just never create historical UserMessage rows for mark-as-read, but still allow marking those with existing UserMessage rows as read?

I'm not sure how much of a refactor this would be.